### ISSUE TITLE
SIGSEGV in ds4_gpu_encode_get_rows_q8_0 when running GLM-5.3-Flash distributed (pipeline parallelism) — crash at first generation step

### ISSUE THREAD

**Issue by dongzhang777-source** [antirez/ds4#901]
**Title: SIGSEGV in ds4_gpu_encode_get_rows_q8_0 when running GLM-5.3-Flash distributed (pipeline parallelism) — crash at first generation step**

## Summary

GLM-5.3-Flash Q2 runs through distributed pipeline prefill successfully, but the coordinator deterministically crashes (EXC_BAD_ACCESS) as soon as generation begins. Reproduced 3× with identical stack. Prefill itself completes without errors and the route handshake is clean, so the failure is specific to the GLM distributed generation path.

## Environment

- Branch `glm-5.3-flash` @ `767e517` (both machines, rebuilt)
- 2× MacBook Pro M1 Max 64GB, macOS 26.6.2, arm64, Metal backend
- `glm53-q2` GGUF (~90 GiB, SHA256 `e81fd6241c6e55a64e1e14e47a3eab61a173fa8d7e4b5c1d1848827119705b32`), same file on both machines (verified after copy)
- Connected point-to-point over a Thunderbolt 4 bridge (static IPs, 0.8 ms RTT)

## Reproduction

Worker (machine B):

```sh
./ds4 --metal -m ~/models/GLM-5.3-Flash-Q2.gguf \
  --role worker --layers 22:output \
  --coordinator 192.168.100.1 5555 -c 16384
```

Coordinator (machine A):

```sh
./ds4 --metal -m /path/to/GLM-5.3-Flash-Q2.gguf \
  --role coordinator --layers 0:21 \
  --listen 192.168.100.1 5555 -c 16384 -n 48 --nothink \
  -p "Introduce yourself in one sentence"
```

Route comes up (`local 0:21 -> <worker> Q2 22:output`), the 16-token prompt prefills to 100%, then the coordinator dies at the first generation step. Also reproduced with `-c 32768`, with `--debug`, and without `--nothink`.

## Crash (coordinator, 3/3 identical)

```
EXC_BAD_ACCESS (SIGSEGV), KERN_INVALID_ADDRESS at 0x3d59c1dfb8d2f2e4
  (note: "possible pointer authentication failure")

Metal                                   MTLResourceListAddResource
AGXMetalG13X                            -[AGXG13XFamilyComputeContext setComputePipelineState:]
ds4                                     ds4_gpu_encode_get_rows_q8_0
ds4                                     ds4_gpu_embed_tokens_q8_0_tensor
ds4                                     glm_graph_forward_indexed_tokens
ds4                                     ds4_session_eval_layer_slice
ds4                                     dist_coordinator_eval_span
ds4                                     ds4_dist_session_eval
```

A second variant crash (different run) aborted in `AutoreleasePoolPage::busted` with "autorelease pool page corrupted", which suggests the buffer holding the token/row arguments may be overrun slightly earlier, on the same code path (`glm_graph_forward_indexed_tokens` → `ds4_gpu_embed_tokens_q8_0_tensor`).

## Variables already ruled out

- Memory pressure (reproduced on a freshly-freed machine, 92% free)
- `--debug` telemetry (on/off)
- Context size (16384 / 32768)
- The worker side: loads and maps its slice fine (47.57 GiB resident, layers 22:44+output), stays healthy through the coordinator's crash

## Guess

`ds4_gpu_encode_get_rows_q8_0` is shared with the DeepSeek path; the GLM 5.3 embedding/n-gram table geometry (row size or alignment for the indexed get-rows) may differ in a way that leaves an argument buffer undersized or an unregistered MTLResource, hence the wild pointer when the encoder adds resources.

Happy to test a fix or provide more diagnostics (`DS4_DIST_CONNECT_TRACE`, full `.ips` files, worker-side logs).

---

**Disclosure:** This issue was drafted by GLM-5.3-Flash (running as the coding agent inside ZCode) at the direction of the machine owner, who reviewed and approved it. All crash reports, stack traces, and reproduction details above were collected first-hand on the owner's hardware.

**Comment by dongzhang777-source** [comment]
## Update after clean reboot (2026-08-30): NOT fixed

Pulled latest `glm-5.3-flash` branch (b1b4ea0, 6 new commits incl. 'Fix GLM 5.3 continued prefill'), rebuilt both machines, and tested on a **freshly rebooted B machine** (GPU fully reset). Generation crash still reproduces 100% (4/4 attempts).

**New finding: the crash point changed.** Original report was SIGSEGV in `ds4_gpu_encode_get_rows_q8_0` → `glm_graph_forward_indexed_tokens` (during generation). Latest branch instead crashes earlier at GPU-driver level:

```
EXC_BAD_ACCESS KERN_INVALID_ADDRESS at 0xbc7e9258bc0f170a
  IOGPU __52-[IOGPUMetalDevice initWithAcceleratorPort:options:
  libdispatch.dylib _dispatch_client_callout / _dispatch_continuation_pop / ...
```

- Environment: 2× M1 Max 64GB, macOS 26.6.2, Metal, ds4 branch b1b4ea0
- A = coordinator layers 0:21, B = worker layers 22:output (GLM 5.3 has 45 layers total), ctx 32768
- Worker loads fine (47.57 GiB resident, KV 0.20 GiB, route ready) — crash is at generation start, reproducible across clean reboot
- Second crash variant seen once: B worker 'Metal model warmup failed: Insufficient Memory' (GPU wedged after repeated failures), but the IOGPU-init crash happens even on first attempt after clean boot

The new code changed the failure mode (no longer `glm_graph_forward_indexed_tokens`) but the underlying generation crash remains. Stack trace attached.

**Comment by dongzhang777-source** [comment]
## Third test on latest main (ab06d19): GLM 5.3 merged to main, crash persists — new stack

GLM 5.3 support has been merged to main (commits including 'Fix GLM streaming resident layer maps', 'Make GLM SSD streaming memory budgets adaptive', etc.). Rebuilt from ab06d19 and tested on freshly rebooted machines. Generation crash still reproduces.

**New crash point (third distinct location):**
```
EXC_BAD_ACCESS
  libCoreFSCache.dylib       fscache_insert_and_retain
  Metal                      MTLCompilerFSCache::addElement(MTLUINT256_t, void const*)
  Metal                      MultiLevelAirCache::addElement(MTLBinaryKey*, MTLAirEntry*)
  Metal                      __130-[_MTLFunctionInternal newSpecializedFunctionWithD...
  Metal                      MTLCompilerScheduler::...
```

**Three crashes across three versions, three different code paths:**
1. Branch 767e517: `ds4_gpu_encode_get_rows_q8_0` → `glm_graph_forward_indexed_tokens` (ds4 compute layer)
2. Branch b1b4ea0: `IOGPUMetalDevice initWithAcceleratorPort` (GPU driver init)
3. Main ab06d19: `MTLCompilerFSCache::addElement` (Metal shader compiler cache)

All EXC_BAD_ACCESS. The changing crash point suggests memory corruption that manifests wherever the corrupted region is next accessed.

**Hypothesis:** GLM 5.3's Metal backend requires extensive kernel specialization (per-layer × per-quant combinations). In distributed mode, two processes (coordinator + worker) trigger simultaneous Metal shader compilations. The FSCache may have a race condition or overflow when both processes compile the same specialized kernels concurrently over the TB bridge.

**Context:** single-machine GLM 5.3 works on 128GB M5 Max (community verified in #868). The issue is specific to distributed/pipeline mode on 2× M1 Max 64GB over Thunderbolt.

Full crash reports and repro steps available in previous comments.

**Comment by dongzhang777-source** [comment]
## Checked latest main (`110afdd`): no relevant changes yet — still blocked

We reviewed the 10 commits between `ab06d19` and `110afdd` (08-31 → 09-01). None touch the GLM distributed (pipeline-parallel) Metal generation path — that batch is DeepSeek Vision-Exp, ROCm/CUDA, and server-focused — so there is no fix to re-run against. The crash still reproduces on `ab06d19` as reported above.

Recapping the impact: on 2×64 GB M1 Max (pipeline parallel over a Thunderbolt Bridge), GLM-5.3-Flash Q2 loads and prefills cleanly on both nodes, but every generation attempt crashes the coordinator at the first generation step — three distinct crash sites so far (`SIGSEGV` in `ds4_gpu_encode_get_rows_q8_0` → `glm_graph_forward_indexed_tokens`; IOGPU wild pointer; `MTLCompilerFSCache::addElement`). Distributed DeepSeek and Qwen run rock-solid on the identical setup (TSO off, wired limits set, clean GPU state), which points to something GLM-distributed-specific rather than the environment.

Happy to retest immediately on any fix branch — the two-machine repro environment and a one-command start script are standing by.

*This comment was drafted by GLM-5.3-Flash (via ZCode) and posted with the account owner's approval.*

**Comment by dongzhang777-source** [comment]
**Follow-up: retested on `main` @ `f62ca29` — the SIGSEGV is fixed, but generation output is now deterministically degenerate**

## Retest setup (2026-09-07)

Same hardware combo: 2× MacBook Pro M1 Max 64GB, pipeline parallelism over a Thunderbolt 4 bridge, GLM-5.3-Flash Q2 (`e81fd624…`, SHA verified on both machines after re-copy).

- Both machines rebuilt from `main` @ `f62ca29` (71 commits after `ab06d19`, including `0e9cc2d` GLM attention masking fix and the Metal buffer-management changes)
- Split `0:21 / 22:output`, ctx 32768, tested via both `ds4` CLI and `ds4-server` (OpenAI API mode)
- One deployment note: the new build now requires `metal/deepseek4_vision.metal` (and the rest of `metal/`) present in the working directory — a stale `metal/` dir on the worker causes `metal backend unavailable; aborting startup`. Syncing `metal/` resolved it.

## Result 1: the crash from this issue is GONE ✅

Prefill completes and generation now runs without any SIGSEGV — no `ds4_gpu_encode_get_rows_q8_0` / `MTLResourceListAddResource` crash across multiple sequential requests. The server also survives; whatever was corrupting the resource list appears fixed.

## Result 2: new symptom — degenerate output ❌

Generation produces the same token repeatedly. Via `ds4-server`, every request returns:

```json
"content": "",
"reasoning_content": "!!!!!!!!!!!!!!!!!!!!!!!!... (repeats until max_tokens)"
```

This is invariant across:

- `temperature` 0 and 1.0, `top_p` 0.9 (so not a sampling artifact — logits appear pinned to one token)
- `enable_thinking: false`
- Different prompts (EN/CN), short and longer
- CLI coordinator and `ds4-server` paths

Both nodes' logs are clean, and the drift-patch flags match exactly on coordinator and worker (`hc_stable=on norm_unify=on kv_raw_f32=off rope_exp2_log2=off math_safe=off tensor_matmul=off`). Worker: `layers 22:44+output, 48 spans, 47.57 GiB`; coordinator: `layers 0:21+output, 46 spans, 40.80 GiB`; route handshake normal; KV allocation normal on both sides.

## Reading

Since the first generated token is already "!" regardless of sampling and prompt, the hidden-state handoff at the PP boundary (or the first coordinator-layer decode step after receiving worker activations) seems to be producing garbage rather than crashing now. Happy to test hypotheses — e.g. a debug knob that dumps worker→coordinator activation checksums, or a forced single-rank run of the same split on one 128GB machine to isolate whether the numerics break only in the distributed path.

---
*This report was drafted with AI assistance (GLM-5.3-Flash via ZCode), reviewed and confirmed by me. All measurements above were run on my hardware.*
